src/ch341a.c

```diff
@@ -0,0 +1,467 @@
+      return ENOMEM;
+    case LIBUSB_ERROR_NOT_SUPPORTED:
+      return ENOSYS;
+    default:
+      return ERANGE;
+  }
+}
+
+#endif
+
+/*
+ * Private data for this programmer.
+ */
+struct pdata
+{
+  libusb_device_handle *usbhandle;
+  int sckfreq_hz;
+};
+
+#define PDATA(pgm) ((struct pdata *)(pgm->cookie))
+
+// ----------------------------------------------------------------------
+/* Prototypes */
+static void ch341a_setup(PROGRAMMER *pgm);
+static void ch341a_teardown(PROGRAMMER *pgm);
+static int ch341a_open(PROGRAMMER *pgm, const char *port);
+static void ch341a_close(PROGRAMMER *pgm);
+static int ch341a_initialize(const PROGRAMMER *pgm, const AVRPART *p);
+static int ch341a_spi_cmd(const PROGRAMMER *pgm, const unsigned char *cmd,
+                          unsigned char *res);
+static int ch341a_spi(const PROGRAMMER *pgm, const unsigned char *in,
+                      unsigned char *out, int size);
+static int ch341a_spi_program_enable(const PROGRAMMER *pgm, const AVRPART *p);
+static int ch341a_spi_chip_erase(const PROGRAMMER *pgm, const AVRPART *p);
+static int ch341a_spi_transfer(const PROGRAMMER *pgm, const unsigned char *cmd,
+                               unsigned char *res);
+// dummy functions
+static void ch341a_disable(const PROGRAMMER *pgm);
+static void ch341a_enable(PROGRAMMER *pgm, const AVRPART *p);
+static void ch341a_display(const PROGRAMMER *pgm, const char *p);
+
+/* ch341 requires LSB first, swap the bit order before send and after receive */
+static unsigned char swap_byte(unsigned char byte)
+{
+  byte = ((byte >> 1) & 0x55) | ((byte << 1) & 0xaa);
+  byte = ((byte >> 2) & 0x33) | ((byte << 2) & 0xcc);
+  byte = ((byte >> 4) & 0x0f) | ((byte << 4) & 0xf0);
+  return byte;
+}
+
+static int CH341USBTransferPart(const PROGRAMMER *pgm,
+                                enum libusb_endpoint_direction dir,
+                                unsigned char *buff, unsigned int size)
+{
+  int ret, bytestransferred;
+
+  if (!PDATA(pgm)->usbhandle) return 0;
+
+  if ((ret = libusb_bulk_transfer(PDATA(pgm)->usbhandle,
+                                  CH341A_USB_BULK_ENDPOINT | dir, buff, size,
+                                  &bytestransferred, CH341A_USB_TIMEOUT))) {
+    pmsg_error("libusb_bulk_transfer for IN_EP failed: %d (%s)\n",
+        ret, libusb_error_name(ret));
+    return -1;
+  }
+  return bytestransferred;
+}
+
+static bool CH341USBTransfer(const PROGRAMMER *pgm,
+                             enum libusb_endpoint_direction dir,
+                             unsigned char *buff, unsigned int size)
+{
+  int pos = 0, bytestransferred;
+  while (size) {
+    bytestransferred = CH341USBTransferPart(pgm, dir, buff + pos, size);
+    if (bytestransferred <= 0) return false;
+    pos += bytestransferred;
+    size -= bytestransferred;
+  }
+  return true;
+}
+
+/* The assumed map between UIO command bits, pins on CH341A chip and pins on SPI chip:
+ * UIO  CH341A  SPI CH341A SPI name
```

Comment:
Please refer to AVR parts in this comment. We use signal names sdi, sdo, sck and reset. This comment isn't really helpful for how to rig up an AVR part so this programmer works with them.